### PR TITLE
Add .editorconfig file

### LINKED ISSUE #586: Add a .editorconfig
for use with http://editorconfig.org/

Some potential contributors struggle with formatting issues, specifically around whitespace.
In some cases, these are easily worked around, and sometimes not.

The presence of a `.editorconfig` file may help avoid some problems. Not everybody is using an editor (or has the plugin) that will respect these settings, but if even a few do, it's no great hardship to add and maintain the file.

Thoughts?

### PR DESCRIPTION
**Pull request by thomaslevesque**
Fix #586

### CODE REVIEW

**Comment by blairconrad** [comment]
https://github.com/FakeItEasy/FakeItEasy/blob/22796eb6f110857d24483f8858fac6960722bcf9/CONTRIBUTING.md#spaces-not-tabs

mentions a 2-space indent for the rakefile. You know, while we're here.

File: Source/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,12 @@
+﻿; Top-most EditorConfig file
```
Comment (blairconrad) [inline Source/.editorconfig]: I'm new to this, but I gotta say—I was excited about seeing

```
# EditorConfig is awesome: http://EditorConfig.org
```

File: Source/.editorconfig [hunk 2]
```diff
@@ -0,0 +1,12 @@
+﻿; Top-most EditorConfig file
+root = true
+
+[*.cs]
+; Windows-style newlines
+end_of_line = crlf
+; 4-column space indentation
```
Comment (blairconrad) [inline Source/.editorconfig]: super nitpicky, but would anybody mind a newline between 6 and 7?

**Comment by blairconrad** [comment]
Generally I like it. Nice job!

File: Source/.editorconfig (hunk 1 above)
Comment (blairconrad) [inline Source/.editorconfig]: :stuck_out_tongue_winking_eye:

**Comment by thomaslevesque** [comment]
> https://github.com/FakeItEasy/FakeItEasy/blob/22796eb6f110857d24483f8858fac6960722bcf9/CONTRIBUTING.md#spaces-not-tabs
> 
> mentions a 2-space indent for the rakefile. You know, while we're here.

It also says that it should be 4 spaces for all file types (except Ruby), so I should probably apply this to `[*]`

File: Source/.editorconfig [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# 4-column space indentation
+indent_style = space
+indent_size = 4
+
+charset = utf-8
+trim_trailing_whitespace = true
+
+[*.cs]
+# Windows-style newlines
+end_of_line = crlf
```
Comment (blairconrad) [inline Source/.editorconfig]: A while back, @adamralph commented about me missing a newline in a file. 
Ah: https://github.com/FakeItEasy/FakeItEasy/pull/566#discussion_r47742056

which was my introduction to .editorconfig. 

Do we want to take that step as well? I've no objection.

Comment (thomaslevesque) [inline Source/.editorconfig]: I don't mind either. Will add it.

**Comment by thomaslevesque** [comment]
I also set the `max_line_length` to 160, as mentioned in the contribution guidelines

File: Source/.editorconfig [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿# EditorConfig is awesome: http://EditorConfig.org
```
Comment (blairconrad) [inline Source/.editorconfig]: Ha! You did it! I was joking. Mostly. I do think it's fun. But I'm quite silly.

Comment (thomaslevesque) [inline Source/.editorconfig]: Well, I guess it can't hurt to include the link, in case someone sees this file and wonders what it is ;)

**Comment by blairconrad** [comment]
Looks like that's only supported by vi and emacs, but shouldn't do any harm. I wouldn't think.

**Comment by blairconrad** [comment]
Looks good to me. I'm for merging, but I've never used editorconfig, so wouldn't mind a chime-in from an experienced user.

Nice job, @thomaslevesque!

File: Source/.editorconfig (hunk 4 above)
Comment (blairconrad) [inline Source/.editorconfig]: True. But even aside from that, it makes me smile.

**Comment by thomaslevesque** [comment]
I just squashed the commits; no point in having 4 distinct commits for something that's new...

**Comment by blairconrad** [comment]
Yeah, I noticed the superfluous commits, but I assumed you were going to squash. Because you're like that. :+1:

File: Source/.editorconfig [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+charset = utf-8
+trim_trailing_whitespace = true
+
+max_line_length = 160
+
+insert_final_newline = true
+
+[*.cs]
+# Windows-style newlines
+end_of_line = crlf
```
Comment (adamralph) [inline Source/.editorconfig]: I would probably move this line into the `[*]` section and kill the `[*.cs]` section. If we ever add any file types which need LF, such as `.sh` then we can add a section for them.

File: Source/.editorconfig [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+[*.rb]
+# 2-column space indentation
+indent_style = space
```
Comment (adamralph) [inline Source/.editorconfig]: This line is redundant and can be removed. It is covered in the `[*]` section.

**Comment by adamralph** [comment]
@thomaslevesque I've reviewed the PR and added a couple of comments. If you agree, I'd be happy if you amended the current commit.

File: Source/.editorconfig (hunk 5 above)
Comment (thomaslevesque) [inline Source/.editorconfig]: OK, it makes sense

**Comment by blairconrad** [comment]
It just occurs to me. Should .editorconfig not have gone at the root level (not in `Source`)?

**Comment by adamralph** [comment]
`.editorconfig` has to be at the same level as the solution in order for the VS add-in to recognise it. 

Incidentally, I think we should move to this layout - https://gist.github.com/davidfowl/ed7564297c61fe9ab814

**Comment by thomaslevesque** [comment]
> Incidentally, I think we should move to this layout - https://gist.github.com/davidfowl/ed7564297c61fe9ab814

I like it! I don't really have an issue with the current layout, but the one you propose looks cleaner

**Comment by blairconrad** [comment]
> .editorconfig has to be at the same level as the solution in order for the VS add-in to recognise it.

Oh? 

> To load the settings for a particular source file, the plugin searches for a file named ".editorconfig" in the same directory as the source file and in each parent directory.

from the [extension's description](https://visualstudiogallery.msdn.microsoft.com/c8bccfe2-650c-4b42-bc5c-845e21f96328).

Not doubting you, but I'd certainly not expected a requirement to be sibling to the .sln.

**Comment by adamralph** [comment]
I was under the impression that it stops walking up the directory path when hits the .sln level but I could be wrong.

**Comment by blairconrad** [comment]
> Incidentally, I think we should move to this layout - https://gist.github.com/davidfowl/ed7564297c61fe9ab814

Perhaps an update to #279?

Also, I don't love the `.sln` at root, at least not without explanation. What's the reasoning? So it's the `.editorconfig`'s friend?

**Comment by blairconrad** [comment]
> I was under the impression that it stops walking up the directory path when hits the .sln level but I could be wrong.

To the lab!

**Comment by blairconrad** [comment]
I'm back. I moved the .editorconfig to the root directory, set the number of spaces per indent to 7, restarted the IDE, and…
the new setting was respected!

**Pull request merged**